doc/design_AsyncNDRegisterAcessor_and_NumericAddressedBackend.dox

```diff
@@ -0,0 +1,158 @@
+  AsyncVariable has been introduced.\ref asyncAccessor_comment_AsyncVariable "(*)"
+  It contains the weak pointer as a private variable and provides member functions to access the object behind
+  the weak pointer.
+  This ensures two points which otherwise would be left to the implementation, which would be error prone:
+  - The weak pointer is locked before accessing the object.
+  - There is a race condition that the locked object could have a nullptr if the shared pointer is currently being
+     destructed while the lock is being acquired. This is checked as well.
+- The void notification queue in the TransferElement base class is set up as a continuation of the data transport queue,
+  such that the read operations of the notification queue trigger the filling of the user buffer.
+
+\anchor asyncAccessor_comment_AsyncVariable (*) There is AsyncVariable as an untyped base class and AsyncVariableImpl,
+  which is templated to a user type. Like this pointers to variables with different user types can be stored in one list.
+
+\subsection AsyncNDRegisterAccessor_usage Interface for implementing backends
+
+While the AsyncNDRegisterAccessor itself is complete and universal, the AsyncAccessorManager and
+AsyncVariable have pure virtual functions which must be provided by the individual
+backend implementations:
+- AsyncVariable::getNumberOfChannels(), AsyncVariable::getNumberOfSamples(), AsyncVariable::getUnit(),
+  AsyncVariable::getDescription() and AsyncVariable::isWriteable() must be provided by deriving from AsyncVariable or
+  AsyncVariableImpl. The information must directly be provided by the backend.
+  The generating code cannot rely on the catalogue here (which might not even be filled yet when the accessor
+  is being created).\ref asyncAccessor_comment_AsyncVariable "(*2)"
+- The 'virtual function template' AsyncAccessorManager::createAsyncVariable() must return an AsyncVariable
+  implementation with the implemented accessor-describing functions from the previous bullet point, such that the
+  accessor manager can create the AsyncNDRegisterAccessor.
+  When createAsyncVariable() returns the AsyncVariable object, the weak pointer to the AsyncNDRegisterAccessor is not
+  set yet. The reason is that the backend specific code should not have the subscription code, which is the same for
+  all backends and would only be code duplication. And it is not sufficient to have the AsyncVariable, which only
+  contains the weak pointer. The calling code also needs the shared pointer to the AsyncNDRegisterAccessor.
+  createAsyncVariable() is used internally by the AsyncAccessorManager, which creates the AsyncNDRegisterAccessor
+  and adds it to the AsyncVariable in the subscription process.
+- AsyncAccessorManager::activate() needs to be implemented in the following way:
+  -# Lock the mutex which protects the list of weak pointer, so no entries can be added or removed
```

Comment:
You lost me again here... Where does the mutex live, and where the weak pointer? I though the weak pointer is part of the AsyncVariable? So is this actually the list of AsyncVariables? For the mutex, writing out the class member name would help.

Reply:
ah yes. I see the confusion. I will have a look at the class description. I think there it is described better.

The mutex protects the list of AsnycVariables in the manager. Whenever you want to access one of those variables, you have to hold the mutex because an un-subscribe could delete it.
So the mutex serves two purposes.
1. protects the existence of the variable.
2. As you have to hold it anyway, it also protects the contents (sending expections, enabling/diabling and sending data do not all come from the same thead).

Reply:
Wow, this was really confusing and wrong. There is no list of weak pointers, there is a list (actually a map) of AsyncVariables.

Reply:
OK, I think I got this one now.